File: src/main/java/org/taktik/icure/entities/PropertyType.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.entities;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import org.taktik.icure.constants.PropertyTypeScope;
import org.taktik.icure.constants.TypedValuesType;
import org.taktik.icure.entities.base.Identifiable;
import org.taktik.icure.entities.base.StoredDocument;

import java.io.Serializable;

@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonIgnoreProperties(ignoreUnknown = true)
public class PropertyType extends StoredDocument implements Cloneable, Serializable, Identifiable<String> {
	private static final long serialVersionUID = 1L;

	protected String identifier;

	protected TypedValuesType type;

	private PropertyTypeScope scope;

	protected boolean unique;

	protected String editor;

	protected boolean localized;


	public PropertyType() {
	}

	public PropertyType(TypedValuesType type, String identifier) {
		this.type = type;
		this.identifier = identifier;
	}

	public PropertyType(TypedValuesType type, PropertyTypeScope scope, String identifier) {
		this.type = type;
		this.scope = scope;
		this.identifier = identifier;
	}

	public String getIdentifier() {
		return identifier;
	}

	public void setIdentifier(String value) {
		this.identifier = value;
	}

	public TypedValuesType getType() {
		return type;
	}

	public void setType(TypedValuesType value) {
		this.type = value;
	}

	public PropertyTypeScope getScope() {
		return scope;
	}

	public void setScope(PropertyTypeScope scope) {
		this.scope = scope;
	}

	public boolean getUnique() {
		return unique;
	}

	public void setUnique(boolean value) {
		this.unique = value;
	}

	public String getEditor() {
		return editor;
	}

	public void setEditor(String value) {
		this.editor = value;
	}

	public boolean getLocalized() {
		return localized;
	}

	public void setLocalized(boolean value) {
		this.localized = value;
	}

	@Override
	public Object clone() throws CloneNotSupportedException {
		return super.clone();
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;

		PropertyType that = (PropertyType) o;

		if (id != null ? !id.equals(that.id) : that.id != null) return false;

		return true;
	}

	@Override
	public int hashCode() {
		return id != null ? id.hashCode() : 0;
	}
}

File: src/main/kotlin/org/taktik/icure/entities/PropertyType.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.entities

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty
import com.github.pozo.KotlinBuilder
import org.taktik.couchdb.entity.Attachment
import org.taktik.icure.constants.PropertyTypeScope
import org.taktik.icure.constants.TypedValuesType
import org.taktik.icure.entities.base.StoredDocument
import org.taktik.icure.entities.embed.RevisionInfo
import org.taktik.icure.utils.DynamicInitializer
import org.taktik.icure.utils.invoke

@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonIgnoreProperties(ignoreUnknown = true)
@KotlinBuilder
data class PropertyType(
        @JsonProperty("_id") override val id: String,
        @JsonProperty("_rev") override val rev: String? = null,
        @JsonProperty("deleted") override val deletionDate: Long? = null,

        val identifier: String? = null,
        val type: TypedValuesType? = null,
        val scope: PropertyTypeScope? = null,
        val unique: Boolean = false,
        val editor: String? = null,
        val localized: Boolean = false,

        @JsonProperty("_attachments") override val attachments: Map<String, Attachment>? = mapOf(),
        @JsonProperty("_revs_info") override val revisionsInfo: List<RevisionInfo>? = listOf(),
        @JsonProperty("_conflicts") override val conflicts: List<String>? = listOf(),
        @JsonProperty("rev_history") override val revHistory: Map<String, String>? = mapOf()

) : StoredDocument {
    companion object : DynamicInitializer<PropertyType> {
        fun with(type: TypedValuesType, scope: PropertyTypeScope, identifier: String) = PropertyType(id = identifier, type = type, scope = scope, identifier = identifier)
    }

    fun merge(other: PropertyType) = PropertyType(args = this.solveConflictsWith(other))
    fun solveConflictsWith(other: PropertyType) = super.solveConflictsWith(other) + mapOf(
            "identifier" to (this.identifier ?: other.identifier),
            "type" to (this.type ?: other.type),
            "scope" to (this.scope ?: other.scope),
            "unique" to (this.unique),
            "editor" to (this.editor ?: other.editor),
            "localized" to (this.localized)
    )

    override fun withIdRev(id: String?, rev: String) = if (id != null) this.copy(id = id, rev = rev) else this.copy(rev = rev)
    override fun withDeletionDate(deletionDate: Long?) = this.copy(deletionDate = deletionDate)
}
